server/extension/extension.go

```diff
@@ -0,0 +1,392 @@
+	v1alpha1 "github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1"
+	"github.com/argoproj/argo-cd/v2/util/settings"
+	"github.com/ghodss/yaml"
+	"github.com/gorilla/mux"
+	log "github.com/sirupsen/logrus"
+	"k8s.io/utils/pointer"
+)
+
+const (
+	URLPrefix                    = "/extensions"
+	HeaderArgoCDApplicationName  = "Argocd-Application-Name"
+	DefaultConnectionTimeout     = 2 * time.Second
+	DefaultKeepAlive             = 15 * time.Second
+	DefaultIdleConnectionTimeout = 60 * time.Second
+	DefaultMaxIdleConnections    = 30
+)
+
+// ExtensionConfigs defines the configurations for all extensions
+// retrieved from Argo CD configmap (argocd-cm).
+type ExtensionConfigs struct {
+	Extensions []ExtensionConfig `json:"extensions"`
+}
+
+// ExtensionConfig defines the configuration for one extension.
+type ExtensionConfig struct {
+	// Name defines the endpoint that will be used to register
+	// the extension route. Mandatory field.
+	Name    string        `json:"name"`
+	Backend BackendConfig `json:"backend"`
+}
+
+// BackendConfig defines the backend service configurations that will
+// be used by an specific extension. An extension can have multiple services
+// associated. This is necessary when Argo CD is managing applications in
+// external clusters. In this case, each cluster may have its own backend
+// service.
+type BackendConfig struct {
+	ProxyConfig
+	Services []ServiceConfig `json:"services"`
+}
+
+// ProxyConfig allows configuring connection behaviour between Argo CD
+// API Server and the backend service.
+type ProxyConfig struct {
+	// ConnectionTimeout is the maximum amount of time a dial to
+	// the extension server will wait for a connect to complete.
+	// Default: 2 seconds
+	ConnectionTimeout time.Duration `json:"connectionTimeout"`
+
+	// KeepAlive specifies the interval between keep-alive probes
+	// for an active network connection between the API server and
+	// the extension server.
+	// Default: 15 seconds
+	KeepAlive time.Duration `json:"keepAlive"`
+
+	// IdleConnectionTimeout is the maximum amount of time an idle
+	// (keep-alive) connection between the API server and the extension
+	// server will remain idle before closing itself.
+	// Default: 60 seconds
+	IdleConnectionTimeout time.Duration `json:"idleConnectionTimeout"`
+
+	// MaxIdleConnections controls the maximum number of idle (keep-alive)
+	// connections between the API server and the extension server.
+	// Default: 30
+	MaxIdleConnections int `json:"maxIdleConnections"`
+}
+
+// ServiceConfig provides the configuration for a backend service.
+type ServiceConfig struct {
+	// URL is the address where the extension backend must be available.
+	// Mandatory field.
+	URL string `json:"url"`
+
+	// ClusterName if provided, will have to match the application
+	// destination name to have requests properly forwarded to this
+	// service URL.
+	ClusterName string `json:"clusterName"`
```

Comment:
Just curious, why name instead of URL?

Reply:
I agree that this is misleading. Thinking about renaming it to just `Cluster` and it could accept the name or the URL.
Sounds good?

Reply:
Done